File: NC78 ReverseList.java
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if(head == null) {
            return null;
        }
        // 当前节点是head，pre是当前节点的前一节点，next是当前节点的下一节点
        ListNode pre = null;
        ListNode next = null;
        // 循环，让当前节点从指向next到指向pre，可以达到反转链表的效果
        while(head != null) {
            // 先用next保存head的下一个节点的信息，保证单链表不会因为失去head节点的原next节点而就此断裂
            next = head.next;
            head.next = pre;
            // head指向pre后，就继续依次反转下一个节点
            pre = head;
            head = next;
        }
        // 直接输出pre就是我们要得到的反转后的链表
        return pre;
    }
}

File: NC78 ReverseList.cpp
// 第一种方法：非递归方法
/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution {
public:
    ListNode* ReverseList(ListNode* pHead) {
        if(pHead == NULL) return NULL; // 注意程序鲁棒性
        
        ListNode* pNode = pHead; // 当前指针
        ListNode* pReverseHead = NULL; // 新链表的头指针
        ListNode* pPrev = NULL; // 当前指针的前一个结点
        
        while(pNode != NULL) {
            ListNode* pNext = pNode->next; // 链断开之前保存断开位置后边的结点
            
            if(pNext==NULL) {
                pReverseHead = pNode;
            }
            
            pNode->next = pPrev; // 链表反转
            pPrev = pNode;
            pNode = pNext;
        }
        return pReverseHead;
        
    }
};


// 第二种方法：递归方法，更巧妙！
class Solution {
public:
    ListNode* ReverseList(ListNode* pHead) {
        // 如果链表为空或者链表中只有一个元素
        if(pHead==NULL || pHead->next==NULL) return pHead;
         
        // 先反转后面的链表，走到链表的末端结点
        ListNode* pReverseNode = ReverseList(pHead->next);
         
        // 再将当前节点设置为后面节点的后续节点
        pHead->next->next=pHead;
        pHead->next=NULL;
         
        return pReverseNode;
         
    }
};